processor_test.go

```diff
@@ -1,28 +1,51 @@
-package mitmengine_test
+package mitmengine
 
 import (
 	"bufio"
 	"fmt"
+	"github.com/cloudflare/mitmengine/loader"
 	"os"
 	"path/filepath"
 	"regexp"
 	"testing"
 
 	ua "github.com/avct/uasurfer"
-	"github.com/cloudflare/mitmengine"
 	"github.com/cloudflare/mitmengine/db"
 	fp "github.com/cloudflare/mitmengine/fputil"
 	"github.com/cloudflare/mitmengine/testutil"
 )
 
-var emptyConfig = mitmengine.Config{}
+var emptyConfig = Config{}
 
-var testConfig = mitmengine.Config{
+var testConfigFile = Config{
 	BrowserFileName:   filepath.Join("testdata", "mitmengine", "browser.txt"),
 	MitmFileName:      filepath.Join("testdata", "mitmengine", "mitm.txt"),
 	BadHeaderFileName: filepath.Join("testdata", "mitmengine", "badheader.txt"),
 }
 
+// This test config tests the Loader interface that is implemented by the S3 struct. Anyone who
+// contributes additional loaders can either add additional testConfigs here and/or write similar
+// unit tests in the loader package.
+var testConfigS3 = Config{
+	BrowserFileName:   "browser.txt",
+	MitmFileName:      "mitm.txt",
+	BadHeaderFileName: "badheader.txt",
+}
+
+func TestMain(m *testing.M) {
+	// Setup for testConfigS3 (because loaer.NewS3Instance returns two values and thus cannot be called in a struct
+	// declaration).
+	// Only run if s3cfg.toml file exists in project root directory or loader directory.
+	s3Instance, err := loader.NewS3Instance("s3cfg.toml")
+	if err != nil {
```

Comment:
Right now this skips _all_ of the processor tests if the `s3cfg.toml` files isn't present, even the ones that test loading the config from a file. Perhaps a better way to do this would be to group the tests based on the config, and only run the configs3 tests if `s3cfg.toml` is present? So basically, something like this to replace the all of the global test objects the `TestMain` and `NewProcessor` functions.
```suggestion
func TestProcessorConfigEmpty(t *testing.T) {
    emptyConfig := Config{}
    t.Run("New", func(t *testing.T) { _, err := NewProcessor(emptyConfig); testutil.Ok(t, err) })
}
func TestProcessorConfigFile (t *testing.T) {
    testConfigFile := Config{
        BrowserFileName:   filepath.Join("testdata", "mitmengine", "browser.txt"),
        MitmFileName:      filepath.Join("testdata", "mitmengine", "mitm.txt"),
        BadHeaderFileName: filepath.Join("testdata", "mitmengine", "badheader.txt"),
    }
    t.Run("New", func(t *testing.T) { _, err := NewProcessor(testConfigFile); testutil.Ok(t, err) })
    t.Run("Check", func(t *testing.T) { _TestProcessorCheck(t, testConfigFile)})
    t.Run("GetByUASignatureBrowser", func(t *testing.T) { _TestProcessorGetByUASignatureBrowser(t, testConfigFile)})
    t.Run("GetByRequestSignatureMitm", func(t *testing.T) { _TestProcessorGetByRequestSignatureMitm(t, testConfigFile)})
}
func TestProcessorConfigS3 (t *testing.T) {
    s3Instance, _ := loader.NewS3Instance("s3cfg.toml")
    if err != nil {
        t.Skip("s3cfg.toml either does not exist in project root directory or loader directory, or was malformed")
    } 
    testConfigS3 := Config{
        BrowserFileName:   "browser.txt",
        MitmFileName:      "mitm.txt",
        BadHeaderFileName: "badheader.txt",
        Loader: s3Instance,
    }
    t.Run("New", func(t *testing.T) { _, err := NewProcessor(testConfigS3); testutil.Ok(t, err) })
    t.Run("Check", func(t *testing.T) { _TestProcessorCheck(t, testConfigS3)})
    t.Run("GetByUASignatureBrowser", func(t *testing.T) { _TestProcessorGetByUASignatureBrowser(t, testConfigS3)})
    t.Run("GetByRequestSignatureMitm", func(t *testing.T) { _TestProcessorGetByRequestSignatureMitm(t, testConfigS3)})
}
```